### DISCUSSION TITLE
Using npm package inside cordova project

### DISCUSSION THREAD

**Discussion by KRC2000** [apache/cordova#386] [Q&A] [answered]
Hi!
I am trying to create an app and I am new to java-script. 
In my app I want to use `openai` npm package, but I can't figure out how to do it, and if it's even possible.
Outside of cordova project I wrote a working javascript that uses `openai` npm package, here how it looks:
```
const { Configuration, OpenAIApi } = require("openai");

const configuration = new Configuration({
  apiKey: token,
});
const openai = new OpenAIApi(configuration);

const completion = await openai.createChatCompletion({
  // Script goes on
```
I ran it using `node` and it worked fine.
Then I tried to implement this script in the cordova project. 
In the root folder of the project I executed `npm install openai`
Then I copied the script inside www/js/index.js

I then built and ran app on my phone using  `cordova build` and `cordova run android`

Something was wrong. I tried to debug the script, and found out that script execution stops on the  
`const { Configuration, OpenAIApi } = require("openai");`

I started researching about how to use packages in the code, import them and etc. and found myself very confused.

I learned that there is a difference between running the script on the web-page and using `node`. If I am not confused, only some packages can be imported to use with web page.
I am trying to figure this out for 2 days, and my head hurts.

Could someone please give me some advice or tips to move in the right direction? I would be really happy

**Comment by breautek** [comment, marked as answer]
Cordova runs in a webview, which is comparable to a browser.

Node, while runs in a JS engine like the browser, is a very different environment with different APIs (e.g. there is no `require` in the browser as that is a Node module concept)

It is *possible* to pull node packages and bundle them for the browser, but Cordova doesn't do this. In fact Cordova is unopinionated, you provide your built JS files and cordova will include them in the app binary. The only requirement is the JS should be able to run in a browser environment. Therefore to include node modules, you'll need to use a bundler tool such as [Webpack](https://webpack.js.org/), [Browserify](https://browserify.org/), or [Rollup](https://rollupjs.org/) to name a few.

However even if you use those tools, the module **must** be written to support a browser environment. I'm not familiar with the `openai` package, but if it's heavily dependent on node APIs, even if you use a bundler, it may not work easily if at all in a browser environment.

If OpenAI framework has native libraries for android, it might be better to create a [cordova plugin](https://cordova.apache.org/docs/en/11.x/guide/hybrid/plugins/index.html) that includes the android native library for your project, and the cordova plugin serves to provide a JS interface to the native library.

**Reply by KRC2000** [reply]
Thank you so much sir. I see now. I guess I will move openai logic somewhere else, outside the app, I will try to make my app connect to the server that will be doing all the dirty work. Thanks alot for the help, have a great day.
